swarm/pss/pss.go

```diff
@@ -92,24 +92,28 @@ type Pss struct {
 	auxAPIs         []rpc.API         // builtins (handshake, test) can add APIs
 
 	// sending and forwarding
-	fwdPool         map[string]*protocols.Peer  // keep track of all peers sitting on the pssmsg routing layer
+	fwdPool         map[string]*protocols.Peer // keep track of all peers sitting on the pssmsg routing layer
+	fwdPoolMu       sync.Mutex
 	fwdCache        map[pssDigest]pssCacheEntry // checksum of unique fields from pssmsg mapped to expiry, cache to determine whether to drop msg
-	cacheTTL        time.Duration               // how long to keep messages in fwdCache (not implemented)
+	fwdCacheMu      sync.Mutex
+	cacheTTL        time.Duration // how long to keep messages in fwdCache (not implemented)
 	msgTTL          time.Duration
 	paddingByteSize int
 
 	// keys and peers
 	pubKeyPool                 map[string]map[whisper.TopicType]*pssPeer // mapping of hex public keys to peer address by topic.
+	pubKeyPoolMu               sync.Mutex
 	symKeyPool                 map[string]map[whisper.TopicType]*pssPeer // mapping of symkeyids to peer address by topic.
-	symKeyDecryptCache         []*string                                 // fast lookup of symkeys recently used for decryption; last used is on top of stack
-	symKeyDecryptCacheCursor   int                                       // modular cursor pointing to last used, wraps on symKeyDecryptCache array
-	symKeyDecryptCacheCapacity int                                       // max amount of symkeys to keep.
+	symKeyPoolMu               sync.Mutex
+	symKeyDecryptCache         []*string // fast lookup of symkeys recently used for decryption; last used is on top of stack
+	symKeyDecryptCacheCursor   int       // modular cursor pointing to last used, wraps on symKeyDecryptCache array
+	symKeyDecryptCacheCapacity int       // max amount of symkeys to keep.
 
 	// message handling
-	handlers map[whisper.TopicType]map[*Handler]bool // topic and version based pss payload handlers. See pss.Handle()
+	handlers   map[whisper.TopicType]map[*Handler]bool // topic and version based pss payload handlers. See pss.Handle()
+	handlersMu sync.Mutex
 
 	// process
-	lock  sync.Mutex
```

Comment:
`lock` has been removed in favour of individual mutexes for all maps. It might be a good idea to transition the code to syncmap, in order to reduce complexity a bit, but for now, there are no races in detected races in PSS.